File: R/s2buildvrt.R
#' s2buildvrt
#' @description Build a vrt with all Sentinel2 bands
#' @details The function build a virtual raster from a Sentinel2 SAFE product.
#'  For now, only L1C and L2a with long name (< 2016/12/06) are recognised.
#'  Output vrt is at 10m resolution.
#' @param infile `character` Full path of the input SAFE folder (alternatively,
#'  full path of the xml file of the product with metadata).
#' @param outfile `character` (optional) Full path (or full name) where the
#'  output vrt file should be created (default: current directory). If a
#'  directory is provided, the file name will be the same of the SAFE input product.
#' @param utmzone `character`(optional) UTM zone of output products (default:
#'  the first one retrieved from input granules). Note that this function
#'  does not perform reprojections: if no granules refer to the specified
#'  UTM zone, no output is created.
#' @return NULL
#'
#' @author Luigi Ranghetti, phD (2017) \email{ranghetti.l@@irea.cnr.it}
#' @note License: GPL 3.0
#' @export
#' @importFrom reticulate import import_builtins py_str

# TODO>
# - add support for L2A compact name
# - call gdalbuildvrt in a way which ensures that the application is found
# - add uspport for relative paths

s2buildvrt <- function(infile,
                       outfile=".",
                       utmzone="") {

  res <- c("10m","20m","60m") # resolutions used

  # If a directory was passed instead of a xml, retrieve xml name automatically
  if (file.info(infile)$isdir) {
      infile_all <- list.files(infile, "\\.xml$",full.names=TRUE)
      infile <- infile_all[grep("(^S2A_.*\\_V[T0-9]+\\_[T0-9]+\\.xml$)|(^MTD\\_MSIL[12][AC]\\.xml$)",basename(infile_all),perl=TRUE)][1]
      if (length(infile_all)==0 | is.na(infile)) {
        stop("Could not find required XML file for S2 data")
      }
  }
  # TODO check existence

  # retrieve UTM zone
  infile_dir = dirname(infile)
  if (utmzone=="") {
    infile_granules <- list.files(file.path(infile_dir,"GRANULE"))
    infile_granules <- sapply(infile_granules,function(x){gsub(".*_T([0-9]{2})[A-Z]{3}_.*","\\1",x)})
    infile_utm_auto <- unique(infile_granules)
    if (length(infile_utm_auto)>1) {
      warning("More than one UTM zone was found in the product; using the first one.")
    }
    utmzone <- infile_utm_auto[1]
  }

  # define basename for output files
  out_prefix <- gsub(".SAFE$","",basename(infile_dir))

  ## Create VRT intermediate files
  infile_level <- gsub(".*L([12][AC]).*","\\1",infile)
  infile_gdalnames <- paste0("SENTINEL2_L",infile_level,":",infile,":",res,":","EPSG_326",utmzone)
  dir.create(vrt_tmpdir <- tempdir(), showWarnings=FALSE)
  vrt01_names <- file.path(vrt_tmpdir,paste0(out_prefix,"_",res,".vrt"))

  # create separate vrt for files
  if (length(vrt01_names) == length(infile_gdalnames)) {
    py <- import_builtins(convert=FALSE)
    sys <- import("sys",convert=FALSE)
    gdal <- import("osgeo",convert=FALSE)$gdal
    for (i in 1:length(vrt01_names)) {
      vrt_bi <- gdal$Open(infile_gdalnames[i])
      writeLines(py_str(vrt_bi$GetMetadata("xml:VRT")[[0]]$encode("utf-8")), vrt01_names[i])
    }
  } else {
    stop("Internal error (this should not happen).")
  }

  # create separate vrt for bands
  if ("10m" %in% res) {
    system(paste0("gdalbuildvrt -b 1 \"",tempdir(),"/",out_prefix,"_b02.vrt\" \"",tempdir(),"/",out_prefix,"_10m.vrt\""),
           intern = Sys.info()["sysname"] == "Windows")
    system(paste0("gdalbuildvrt -b 2 \"",tempdir(),"/",out_prefix,"_b03.vrt\" \"",tempdir(),"/",out_prefix,"_10m.vrt\""),
           intern = Sys.info()["sysname"] == "Windows")
    system(paste0("gdalbuildvrt -b 3 \"",tempdir(),"/",out_prefix,"_b04.vrt\" \"",tempdir(),"/",out_prefix,"_10m.vrt\""),
           intern = Sys.info()["sysname"] == "Windows")
    system(paste0("gdalbuildvrt -b 4 \"",tempdir(),"/",out_prefix,"_b08.vrt\" \"",tempdir(),"/",out_prefix,"_10m.vrt\""),
           intern = Sys.info()["sysname"] == "Windows")
  }
  if ("20m" %in% res) {
    system(paste0("gdalbuildvrt -b 1 \"",tempdir(),"/",out_prefix,"_b05.vrt\" \"",tempdir(),"/",out_prefix,"_20m.vrt\""),
           intern = Sys.info()["sysname"] == "Windows")
    system(paste0("gdalbuildvrt -b 2 \"",tempdir(),"/",out_prefix,"_b06.vrt\" \"",tempdir(),"/",out_prefix,"_20m.vrt\""),
           intern = Sys.info()["sysname"] == "Windows")
    system(paste0("gdalbuildvrt -b 3 \"",tempdir(),"/",out_prefix,"_b07.vrt\" \"",tempdir(),"/",out_prefix,"_20m.vrt\""),
           intern = Sys.info()["sysname"] == "Windows")
    system(paste0("gdalbuildvrt -b 4 \"",tempdir(),"/",out_prefix,"_b08a.vrt\" \"",tempdir(),"/",out_prefix,"_20m.vrt\""),
           intern = Sys.info()["sysname"] == "Windows")
    system(paste0("gdalbuildvrt -b 5 \"",tempdir(),"/",out_prefix,"_b11.vrt\" \"",tempdir(),"/",out_prefix,"_20m.vrt\""),
           intern = Sys.info()["sysname"] == "Windows")
    system(paste0("gdalbuildvrt -b 6 \"",tempdir(),"/",out_prefix,"_b12.vrt\" \"",tempdir(),"/",out_prefix,"_20m.vrt\""),
           intern = Sys.info()["sysname"] == "Windows")
  }
  if ("60m" %in% res) {
    system(paste0("gdalbuildvrt -b 1 \"",tempdir(),"/",out_prefix,"_b01.vrt\" \"",tempdir(),"/",out_prefix,"_60m.vrt\""),
           intern = Sys.info()["sysname"] == "Windows")
    system(paste0("gdalbuildvrt -b 2 \"",tempdir(),"/",out_prefix,"_b09.vrt\" \"",tempdir(),"/",out_prefix,"_60m.vrt\""),
           intern = Sys.info()["sysname"] == "Windows")
    system(paste0("gdalbuildvrt -b 3 \"",tempdir(),"/",out_prefix,"_b10.vrt\" \"",tempdir(),"/",out_prefix,"_60m.vrt\""),
           intern = Sys.info()["sysname"] == "Windows")
  }

  # create final vrt
  if (file.info(outfile)$isdir) {
    outfile <- file.path(outfile,paste0(out_prefix,".vrt"))
  }
  system(
    paste0(
      "gdalbuildvrt -separate ",
      "-resolution highest ",
      "\"",outfile,"\" ",
      if ("60m" %in% res) {paste0("\"",tempdir(),"/",out_prefix,"_b01.vrt\" ")},
      if ("10m" %in% res) {paste0("\"",tempdir(),"/",out_prefix,"_b02.vrt\" ")},
      if ("10m" %in% res) {paste0("\"",tempdir(),"/",out_prefix,"_b03.vrt\" ")},
      if ("10m" %in% res) {paste0("\"",tempdir(),"/",out_prefix,"_b04.vrt\" ")},
      if ("20m" %in% res) {paste0("\"",tempdir(),"/",out_prefix,"_b05.vrt\" ")},
      if ("20m" %in% res) {paste0("\"",tempdir(),"/",out_prefix,"_b06.vrt\" ")},
      if ("20m" %in% res) {paste0("\"",tempdir(),"/",out_prefix,"_b07.vrt\" ")},
      if ("10m" %in% res) {paste0("\"",tempdir(),"/",out_prefix,"_b08.vrt\" ")},
      if ("20m" %in% res) {paste0("\"",tempdir(),"/",out_prefix,"_b08a.vrt\" ")},
      if ("60m" %in% res) {paste0("\"",tempdir(),"/",out_prefix,"_b09.vrt\" ")},
      if ("60m" %in% res) {paste0("\"",tempdir(),"/",out_prefix,"_b10.vrt\" ")},
      if ("20m" %in% res) {paste0("\"",tempdir(),"/",out_prefix,"_b11.vrt\" ")},
      if ("20m" %in% res) {paste0("\"",tempdir(),"/",out_prefix,"_b12.vrt\" ")}
    ), intern = Sys.info()["sysname"] == "Windows"
  )

}

File: inst/s2download/s2buildvrt.py
#!/usr/bin/env python
"""
Create a vrt with all the S2 bands.
DEPRECATED: use the R function s2_translate.R in package RSPrePro instead.

Author: Luigi Ranghetti

Year: 2017

License: GPS 3.0
"""

import os
import sys
from osgeo import gdal
from osgeo import osr
import glob
import re
import subprocess
import argparse
import warnings
import tempfile


# Check GDAL version >= 2.1.3
# (2.1.0 is sufficient for long naming convention, but 2.1.3 is required
# since 2016-12-06 to recognise compact names)
if gdal.__version__ < "2.1.2":
    print("This script requires GDAL 2.1.3 or newer")
    sys.exit(1)
        
def s2buildvrt(infile,outdir,infile_utm):
       
    # If a directory were passed instead of a xml, retrieve xml name automatically
    if os.path.isdir(infile):
        try:
            infile_all = glob.glob(os.path.join(os.path.abspath(infile), "*.xml"))
            infile = [f for f in infile_all if re.search('(^S2A_.*\_V[T0-9]+\_[T0-9]+\.xml$)|(^MTD\_MSIL[12][AC]\.xml$)', os.path.basename(f))][0]
        except IndexError:
            raise IOError("Could not find required XML file for S2 data")
    else:
        infile = os.path.abspath(infile)
    # TODO check existence
        
    # retrieve UTM zone
    if infile_utm=="":
        infile_dir = os.path.dirname(infile)
        infile_granules = [os.path.basename(p) for p in glob.glob(os.path.join(infile_dir,"GRANULE","*"))]
        infile_utm_auto = list(set([re.search('\_T([0-9]{2})[A-Z]{3}\_',g).group(1) for g in infile_granules]))
        if len(infile_utm_auto)>1:
            warnings.warn("More than one UTM zone was found in the product; using the first one.", UserWarning)
        infile_utm = infile_utm_auto[0]
    
    # create separate vrt for files
    infile_level = re.search('L([12][AC])', infile).group(1)
    infile_gdalname = ["SENTINEL2_L"+infile_level + ":" + infile + ":" + res + ":" + "EPSG_326"+infile_utm for res in ["10m","20m","60m","TCI"]]
    resvrt_name = [os.path.join(tempfile.gettempdir,"tmp01_"+res+".vrt") for res in ["10m","20m","60m","TCI"]]
    infile_gdal = [gdal.Open(i) for i in infile_gdalname]
    for sel_resvrt_name,sel_infile_gdal in zip(resvrt_name,infile_gdal):
        open(sel_resvrt_name, 'wb').write(sel_infile_gdal.GetMetadata('xml:VRT')[0].encode('utf-8'))




    #SENTINEL2_L1C:MTD_MSIL1C.xml:10m:EPSG_32632 10m.vrt


"""
python2 -c "import sys; from osgeo import gdal; ds = gdal.Open(sys.argv[1]); open(sys.argv[2], 'wb').write(ds.GetMetadata('xml:VRT')[0].encode('utf-8'))" SENTINEL2_L1C:MTD_MSIL1C.xml:10m:EPSG_32632 10m.vrt
python2 -c "import sys; from osgeo import gdal; ds = gdal.Open(sys.argv[1]); open(sys.argv[2], 'wb').write(ds.GetMetadata('xml:VRT')[0].encode('utf-8'))" SENTINEL2_L1C:MTD_MSIL1C.xml:20m:EPSG_32632 20m.vrt
python2 -c "import sys; from osgeo import gdal; ds = gdal.Open(sys.argv[1]); open(sys.argv[2], 'wb').write(ds.GetMetadata('xml:VRT')[0].encode('utf-8'))" SENTINEL2_L1C:MTD_MSIL1C.xml:60m:EPSG_32632 60m.vrt

gdalbuildvrt -b 1 b01.vrt 60m.vrt
gdalbuildvrt -b 1 b01.vrt 60m.vrt
gdalbuildvrt -b 1 b02.vrt 10m.vrt
gdalbuildvrt -b 2 b03.vrt 10m.vrt
gdalbuildvrt -b 3 b04.vrt 10m.vrt
gdalbuildvrt -b 1 b05.vrt 20m.vrt
gdalbuildvrt -b 2 b06.vrt 20m.vrt
gdalbuildvrt -b 3 b07.vrt 20m.vrt
gdalbuildvrt -b 4 b08.vrt 10m.vrt
gdalbuildvrt -b 4 b08a.vrt 20m.vrt
gdalbuildvrt -b 2 b09.vrt 60m.vrt
gdalbuildvrt -b 3 b10.vrt 60m.vrt
gdalbuildvrt -b 5 b11.vrt 20m.vrt
gdalbuildvrt -b 6 b12.vrt 20m.vrt

gdalbuildvrt -separate -resolution highest l1c.vrt b01.vrt b02.vrt b03.vrt 
"""




if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Create a vrt with all the S2 bands.")
    parser.add_argument("infile",nargs="+",
                        help="Input file")
    parser.add_argument("-o", "--outdir",
                        help="Output directory",
                        default="./")
    parser.add_argument("-t", "--utm",
                        help="UTM tile",
                        default="")
    args = parser.parse_args()
    
    s2buildvrt(args.infile[0], args.outdir, args.utm)
